File: Private/MergeSort.py
def merge_sort(arr):

  if len(arr) <= 1:
    return arr

  mid = len(arr) // 2
  left = arr[:mid]
  right = arr[mid:]

  return merge_arrays(merge_sort(left), merge_sort(right))

def merge_arrays(a, b):

  merged = []
  while len(a) > 0 and len(b) > 0:
    
    if a[0] < b[0]:
      merged.append(a[0])
      a.pop(0)
    else: 
      merged.append(b[0])
      b.pop(0)
  
  merged += a + b
  return merged



print(merge_sort([1,4,2,3,7,6,5]))

File: Private/merge-sort/merge-sort.js
arr = [3, 4, 8, 13, 1, 5, 6, 11];

const mergeSort = (arr) => {
  console.log('array: ' + arr);
  let n = arr.length;
  let mid = Math.floor(n / 2);
  let left = arr.slice(0, mid);
  let right = arr.slice(mid);
  //   console.log('mid: ' + mid);
  //   console.log('left: ' + left);
  //   console.log('right: ' + right);
  if (n <= 1) {
    return arr;
  }
  return merge(mergeSort(left), mergeSort(right));
}

//merge sorted areas
const merge = (left, right) => {
  let result = [];
  while (left.length > 0 && right.length > 0) {
    if (left[0] <= right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }
  while (left.length > 0) {
    result.push(left.shift());
  }
  while (right.length > 0) {
    result.push(right.shift());
  }
  return result;
}

console.log(mergeSort(arr));
